// jest-ts/dist/transformer.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/transformer.ts
var transformer_exports = {};
__export(transformer_exports, {
  default: () => transformer_default,
  transformTsCode: () => transformTsCode,
  transformer: () => transformer
});
module.exports = __toCommonJS(transformer_exports);
var esbuild = __toESM(require("esbuild"));
var import_fs_extra = __toESM(require("fs-extra"));
var import_path = __toESM(require("path"));
var isObject = (value) => {
  return typeof value === "object" && value !== null;
};
var isValidTransformResult = (result) => {
  return isObject(result) && typeof result.code === "string";
};
var DEFAULT_CACHE_DIR = "<rootDir>/.jest-ts-cache";
var resolveCacheDir = (cacheDir, rootDir) => {
  return cacheDir.replace("<rootDir>", rootDir);
};
var readCacheSync = (filePath, cacheDir) => {
  try {
    const cacheFile = import_path.default.join(cacheDir, `${import_path.default.basename(filePath)}.json`);
    if (!import_fs_extra.default.pathExistsSync(cacheFile)) return null;
    const fileStat = import_fs_extra.default.statSync(filePath);
    const cacheStat = import_fs_extra.default.statSync(cacheFile);
    if (fileStat.mtimeMs > cacheStat.mtimeMs) {
      import_fs_extra.default.removeSync(cacheFile);
      return null;
    }
    const cacheData = import_fs_extra.default.readJsonSync(cacheFile);
    return isValidTransformResult(cacheData) ? { code: cacheData.code, map: cacheData.map } : null;
  } catch (error) {
    console.warn(`[jest-ts] \u540C\u6B65\u7F13\u5B58\u8BFB\u53D6\u5931\u8D25 (${filePath}):`, error);
    return null;
  }
};
var writeCacheSync = (filePath, cacheDir, data) => {
  try {
    import_fs_extra.default.ensureDirSync(cacheDir);
    const cacheFile = import_path.default.join(cacheDir, `${import_path.default.basename(filePath)}.json`);
    import_fs_extra.default.writeJsonSync(cacheFile, { ...data, mtime: Date.now() });
  } catch (error) {
    console.warn(`[jest-ts] \u540C\u6B65\u7F13\u5B58\u5199\u5165\u5931\u8D25 (${filePath}):`, error);
  }
};
var transformTsCodeSync = (sourceText, sourcePath, options, rootDir) => {
  if (!sourcePath.endsWith(".ts") && !sourcePath.endsWith(".tsx")) {
    return { code: sourceText };
  }
  const finalOptions = {
    cacheDirectory: resolveCacheDir(DEFAULT_CACHE_DIR, rootDir),
    typeCheck: false,
    tsconfigPath: import_path.default.resolve(rootDir, "tsconfig.json"),
    target: "node16",
    format: "cjs",
    esbuildOptions: {},
    ...options
  };
  const cachedResult = readCacheSync(sourcePath, finalOptions.cacheDirectory);
  if (cachedResult) {
    return cachedResult;
  }
  try {
    const esbuildResult = esbuild.transformSync(sourceText, {
      loader: sourcePath.endsWith(".tsx") ? "tsx" : "ts",
      target: finalOptions.target,
      format: finalOptions.format,
      sourcemap: "inline",
      banner: '"use strict;"',
      ...finalOptions.esbuildOptions
    });
    const transformResult = {
      code: esbuildResult.code,
      map: esbuildResult.map || void 0
    };
    if (isValidTransformResult(transformResult)) {
      writeCacheSync(sourcePath, finalOptions.cacheDirectory, transformResult);
      return transformResult;
    }
  } catch (error) {
    console.error(`[jest-ts] \u540C\u6B65\u8F6C\u8BD1\u5931\u8D25 (${sourcePath}):`, error);
  }
  return { code: sourceText };
};
var transformTsCode = async (sourceText, sourcePath, options, rootDir) => {
  return new Promise((resolve) => {
    resolve(transformTsCodeSync(sourceText, sourcePath, options, rootDir));
  });
};
var transformer = {
  /**
   * 同步处理方法（Jest 30.x 强制要求）
   */
  process(sourceText, sourcePath, options) {
    const rootDir = options.config.rootDir;
    const userOptions = options.globals?.["jest-ts"] || {};
    const result = transformTsCodeSync(sourceText, sourcePath, userOptions, rootDir);
    if (!isValidTransformResult(result)) {
      console.error(`[jest-ts] \u540C\u6B65\u8F6C\u8BD1\u7ED3\u679C\u65E0\u6548 (${sourcePath})`);
      return { code: sourceText };
    }
    return result;
  },
  /**
   * 异步处理方法（推荐使用）
   */
  processAsync: async (sourceText, sourcePath, options) => {
    return transformTsCode(sourceText, sourcePath, options.globals?.["jest-ts"] || {}, options.config.rootDir);
  },
  // 版本标识
  version: "1.0.0"
};
var transformer_default = transformer;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  transformTsCode,
  transformer
});
