// jest-ts/src/cache.ts
import { existsSync, readFileSync, writeFileSync, mkdirSync } from "fs";
import { resolve, dirname } from "path";
import { createHash } from "crypto";
import type { JestTsOptions, CacheData } from "./types";

export class CacheManager {
  private readonly cacheDir: string;
  private readonly version: string;

  constructor(config: JestTsOptions, version: string) {
    this.cacheDir = config.cacheDirectory;
    this.version = version;
    this.ensureCacheDir();
  }

  private ensureCacheDir(): void {
    if (!existsSync(this.cacheDir)) {
      mkdirSync(this.cacheDir, { recursive: true });
    }
  }

  getConfigHash(config: JestTsOptions): string {
    return createHash("md5")
      .update(JSON.stringify(config) + this.version)
      .digest("hex");
  }

  getCacheKey(content: string, filename: string, configHash: string): string {
    return createHash("md5")
      .update(content + filename + configHash)
      .digest("hex");
  }

  readCache(cacheKey: string): CacheData | null {
    const cachePath = resolve(this.cacheDir, `${cacheKey}.json`);
    if (!existsSync(cachePath)) return null;

    try {
      return JSON.parse(readFileSync(cachePath, "utf8")) as CacheData;
    } catch {
      return null;
    }
  }

  writeCache(cacheKey: string, data: Omit<CacheData, "mtime">): void {
    const cachePath = resolve(this.cacheDir, `${cacheKey}.json`);
    mkdirSync(dirname(cachePath), { recursive: true });
    writeFileSync(
      cachePath,
      JSON.stringify({ ...data, mtime: Date.now() } as CacheData),
      "utf8"
    );
  }
}
