// jest-ts/src/test.js
const { transformTsCode, CacheManager, resolveConfig } = require('../dist/index');
const fs = require('fs-extra');
const path = require('path');
const { version } = require('../package.json');

// 测试环境准备
const TEST_ROOT = path.join(__dirname, 'test-env');
const TS_FILE = path.join(TEST_ROOT, 'sample.ts');
const CACHE_DIR = path.join(TEST_ROOT, '.jest-ts-cache');
const TS_CONFIG = path.join(__dirname, '../tsconfig.json');

// 初始化测试文件
beforeAll(() => {
  fs.ensureDirSync(TEST_ROOT);
  fs.writeFileSync(TS_FILE, `
    export interface User {
      name: string;
      age: number;
    }
    export const getUser = (): User => ({ name: 'test', age: 20 });
  `, 'utf8');
});

// 清理测试环境
afterAll(() => {
  fs.removeSync(TEST_ROOT);
});

describe('核心功能测试', () => {
  // 测试1：基础转译功能
  test('应正确转译TS代码并移除类型定义', () => {
    const tsCode = `
      const greet = (name: string): string => \`Hello, \${name}!\`;
      export default greet;
    `;

    const result = transformTsCode(TS_FILE, tsCode, {
      'jest-ts': { cacheDirectory: CACHE_DIR, typeCheck: false }
    });

    // 验证类型注解被移除
    expect(result.code).not.toContain(': string');
    expect(result.code).toContain('const greet = (name) => `Hello, ${name}!`;');
    expect(result.code).toContain('export default greet;');
  });

  // 测试2：缓存机制
  test('相同输入应命中缓存，避免重复转译', () => {
    const tsCode = `export const PI = 3.1415926;`;
    const config = resolveConfig({ 'jest-ts': { cacheDirectory: CACHE_DIR } });
    const cacheManager = new CacheManager(config, version);
    
    // 首次转译（无缓存）
    const firstResult = transformTsCode(TS_FILE, tsCode, { 'jest-ts': config });
    const cacheKey = cacheManager.getCacheKey(tsCode, TS_FILE, cacheManager.getConfigHash(config));
    
    // 验证缓存已生成
    expect(cacheManager.readCache(cacheKey)).toEqual(firstResult);
    
    // 二次转译（应命中缓存）
    const secondResult = transformTsCode(TS_FILE, tsCode, { 'jest-ts': config });
    expect(secondResult).toEqual(firstResult);
  });

  // 测试3：类型检查功能
  test('类型错误代码应触发检查失败', () => {
    const invalidCode = `
      const num: number = 'not a number';
      export default num;
    `;

    // 执行带类型检查的转译
    const runCheck = () => transformTsCode(TS_FILE, invalidCode, {
      'jest-ts': {
        cacheDirectory: CACHE_DIR,
        typeCheck: true,
        tsconfigPath: TS_CONFIG
      }
    });

    // 验证错误捕获
    expect(runCheck).toThrow();
    expect(runCheck).toThrow('Type \'string\' is not assignable to type \'number\'');
  });

  // 测试4：配置解析
  test('用户配置应覆盖默认配置', () => {
    const userConfig = {
      'jest-ts': {
        cacheDirectory: './custom-cache',
        typeCheck: false,
        tsconfigPath: './custom-tsconfig.json'
      }
    };

    const resolved = resolveConfig(userConfig);
    
    // 验证覆盖结果
    expect(resolved.cacheDirectory).toBe('./custom-cache');
    expect(resolved.typeCheck).toBe(false);
    expect(resolved.tsconfigPath).toBe('./custom-tsconfig.json');
  });
});
