// jest-ts/dist/cli/init.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/cli/init.ts
var import_fs_extra = __toESM(require("fs-extra"));
var import_path = __toESM(require("path"));
var import_color_printf = require("color-printf");
var filesToGenerate = [
  {
    path: "tsconfig.json",
    content: `{
  "compilerOptions": {
    "target": "ES2020",
    "module": "CommonJS", // \u4E0E\u8F6C\u8BD1\u5668\u8F93\u51FA\u683C\u5F0F\u4E00\u81F4
    "moduleResolution": "Node",
    "strict": true,
    "esModuleInterop": true, // \u517C\u5BB9CommonJS\u548CES\u6A21\u5757
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "outDir": "dist",
    "rootDir": "."
  },
  "include": ["src/**/*", "__tests__/**/*"],
  "exclude": ["node_modules"]
}
`
  },
  {
    path: "src/__test__/math.ts",
    content: `export const add = (a: number, b: number): number => {
  if (typeof a !== 'number' || typeof b !== 'number') {
    throw new TypeError('\u53C2\u6570\u5FC5\u987B\u4E3A\u6570\u5B57');
  }
  return a + b;
};

export const multiply = (a: number, b: number): number => {
  if (typeof a !== 'number' || typeof b !== 'number') {
    throw new TypeError('\u53C2\u6570\u5FC5\u987B\u4E3A\u6570\u5B57');
  }
  return a * b;
};
`
  },
  {
    path: "__tests__/math.test.ts",
    content: `import { add, multiply } from '../src/__test__/math';

describe('src/__test__/math.ts', () => {
  describe('add\u51FD\u6570', () => {
    it('\u8BA1\u7B97\u4E24\u4E2A\u6B63\u6574\u6570\u7684\u548C', () => {
      expect(add(2, 3)).toBe(5);
    });
    it('\u8BA1\u7B97\u8D1F\u6570\u4E0E\u6B63\u6570\u7684\u548C', () => {
      expect(add(-1, 5)).toBe(4);
    });
    it('\u975E\u6570\u5B57\u53C2\u6570\u629B\u51FATypeError', () => {
      expect(() => add(1, '2')).toThrow(TypeError);
    });
  });

  describe('multiply\u51FD\u6570', () => {
    it('\u8BA1\u7B97\u4E24\u4E2A\u6570\u7684\u4E58\u79EF', () => {
      expect(multiply(4, 5)).toBe(20);
    });
    it('\u975E\u6570\u5B57\u53C2\u6570\u629B\u51FATypeError', () => {
      expect(() => multiply(2, null)).toThrow(TypeError);
    });
  });
});
`
  }
];
var generateJestConfig = (projectRoot) => ({
  path: "jest.config.js",
  content: `const path = require('path');

/** @type {import('jest').Config} */
module.exports = {
  // \u786E\u4FDD\u6240\u6709TypeScript\u6587\u4EF6\u90FD\u88AB\u8F6C\u8BD1
  transform: {
    '^.+\\.(ts|tsx)$': path.resolve(
      '${projectRoot}',
      'node_modules/jest-ts/dist/transformer.js'
    )
  },
  // \u660E\u786E\u6307\u5B9A\u6587\u4EF6\u6269\u5C55\u540D\u4F18\u5148\u7EA7
  moduleFileExtensions: ['ts', 'tsx', 'js', 'json', 'node'],
  preset: undefined,
  testEnvironment: 'node',
  testMatch: [
    '**/__tests__/**/*.+(ts|tsx|js)',
    '**/?(*.)+(spec|test).+(ts|tsx|js)'
  ],
  // \u4E0D\u5FFD\u7565\u4EFB\u4F55\u9700\u8981\u8F6C\u8BD1\u7684\u6587\u4EF6
  transformIgnorePatterns: []
};
`
});
var generatePackageJson = async (projectRoot) => {
  const pkgPath = import_path.default.resolve(projectRoot, "package.json");
  const requiredDevDeps = {
    "jest-ts": "^1.0.0",
    "jest": "^29.0.0",
    "typescript": "^5.0.0",
    "esbuild": "^0.20.0"
  };
  try {
    let pkgContent = {};
    if (await import_fs_extra.default.pathExists(pkgPath)) {
      pkgContent = await import_fs_extra.default.readJson(pkgPath, { throws: false }) || {};
      pkgContent.scripts = pkgContent.scripts || {};
      pkgContent.devDependencies = pkgContent.devDependencies || {};
      pkgContent.scripts.test = "jest";
      pkgContent.scripts["test:watch"] = "jest --watch";
      Object.entries(requiredDevDeps).forEach(([dep, version]) => {
        if (!pkgContent.devDependencies[dep]) {
          pkgContent.devDependencies[dep] = version;
        }
      });
      await import_fs_extra.default.writeJson(pkgPath, pkgContent, { spaces: 2 });
    } else {
      pkgContent = {
        name: import_path.default.basename(projectRoot),
        version: "1.0.0",
        private: true,
        scripts: { test: "jest", "test:watch": "jest --watch" },
        devDependencies: requiredDevDeps
      };
      await import_fs_extra.default.writeJson(pkgPath, pkgContent, { spaces: 2 });
    }
    console.log((0, import_color_printf.green)("\u2705 \u5DF2\u914D\u7F6Epackage.json"));
  } catch (error) {
    throw new Error(`\u5904\u7406package.json\u5931\u8D25\uFF1A${error.message}`);
  }
};
var generateSingleFile = async (filePath, fileContent) => {
  const fullPath = import_path.default.resolve(filePath);
  await import_fs_extra.default.ensureDir(import_path.default.dirname(fullPath));
  if (await import_fs_extra.default.pathExists(fullPath)) {
    const overwrite = await promptUserConfirm(`\u6587\u4EF6 ${(0, import_color_printf.yellow)(filePath)} \u5DF2\u5B58\u5728\uFF0C\u662F\u5426\u8986\u76D6\uFF1F(y/N) `);
    if (!overwrite) {
      console.log((0, import_color_printf.gray)(`  \u27A4 \u8DF3\u8FC7\u751F\u6210\uFF1A${filePath}`));
      return;
    }
  }
  await import_fs_extra.default.writeFile(fullPath, fileContent, { encoding: "utf8" });
  console.log((0, import_color_printf.green)(`  \u2705 \u751F\u6210\uFF1A${filePath}`));
};
var promptUserConfirm = (promptText) => {
  return new Promise((resolve) => {
    const readline = require("readline").createInterface({
      input: process.stdin,
      output: process.stdout,
      terminal: process.stdout.isTTY
    });
    readline.question(promptText, (answer) => {
      readline.close();
      resolve(answer.trim().toLowerCase() === "y");
    });
    if (!process.stdout.isTTY) {
      readline.close();
      resolve(false);
    }
  });
};
var init = async () => {
  const projectRoot = process.cwd();
  console.log((0, import_color_printf.blue)(`
\u{1F680} \u521D\u59CB\u5316jest-ts\u6D4B\u8BD5\u73AF\u5883\uFF08\u9879\u76EE\u6839\u76EE\u5F55\uFF1A${projectRoot}\uFF09
`));
  try {
    const transformerPath = import_path.default.resolve(
      projectRoot,
      "node_modules/jest-ts/dist/transformer.js"
    );
    if (!await import_fs_extra.default.pathExists(transformerPath)) {
      throw new Error(`\u274C \u8F6C\u8BD1\u5668\u7F3A\u5931\uFF1A${transformerPath}
\u8BF7\u91CD\u65B0\u5B89\u88C5jest-ts`);
    }
    console.log((0, import_color_printf.gray)("1/2\uFF1A\u751F\u6210\u914D\u7F6E\u6587\u4EF6\u548C\u793A\u4F8B\u4EE3\u7801"));
    const jestConfig = generateJestConfig(projectRoot);
    for (const file of [...filesToGenerate, jestConfig]) {
      await generateSingleFile(file.path, file.content);
    }
    console.log((0, import_color_printf.gray)("\n2/2\uFF1A\u914D\u7F6Epackage.json"));
    await generatePackageJson(projectRoot);
    console.log((0, import_color_printf.blue)(`
\u{1F389} \u521D\u59CB\u5316\u5B8C\u6210\uFF01
`));
    console.log((0, import_color_printf.gray)("1. \u5B89\u88C5\u4F9D\u8D56\uFF1A"));
    console.log(`   ${(0, import_color_printf.cyan)("npm install")}
`);
    console.log((0, import_color_printf.gray)("2. \u8FD0\u884C\u6D4B\u8BD5\uFF1A"));
    console.log(`   ${(0, import_color_printf.cyan)("npm test")}
`);
  } catch (error) {
    console.error((0, import_color_printf.red)(`
\u274C \u521D\u59CB\u5316\u5931\u8D25\uFF1A${error.message}`));
    process.exit(1);
  }
};
init();
